text/0069-api-v2.md

```diff
@@ -0,0 +1,225 @@
+2. `k{keyspace prefix id}x`: TxnKV key.
+3. `k{keyspace prefix id}r`: RawKV key.
+4. `k{keyspace prefix id}a`: RawKV atomic key, on which you can use `CompareAndSwap`.
+
+The `{keyspace prefix id}` is the [keyspace](https://github.com/tikv/rfcs/pull/39) prefix for seperating keys of different keyspace. It should be an vary-length integer whose highest bit of every byte denotes whether the next byte is still part of the integer. The client will fetch the prefix from PD by the keyspace name specified by the user when initializing the client, so it means that the keyspace prefix is valid during the session, in other words, the change on keyspace on PD will not take effect on running seesions.
+
+Note that in TxnKV, the key will be encoded by `Memory Comparable Encoding`. But since the `Memory Comparable Encoding` will not change the starting bytes but only add paddings, there won't be any overlap between RawKV and TxnKV.
+
+#### RawKV Value Encode
+
+If the key has RawKV prefix, which are `k{keyspace id prefix}r` or `k{keyspace id prefix}a`, then the value can be either:
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+#### TTL in Titan
+
+Titan is a storage engine plugin for RocksDB that store the value of a kvpair in a separated blob file if the value is large enough. This brings a new challenge for RawKV TTL compaction filter where the compaction filter can not know the ttl without reading the blob file. So with Titan, the RawKV value encode is defined to be:
+
+##### Value
+
+1. `{0x0}`
+2. `{0x1}{TTL expire timestamp}`
+
+##### Blob Value
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+### Keyspace Management
+
+Add a new http interface to PD for adding, renaming, deleteing and querying the mapping from keyspace name to prefix id:
+
+```javascript
+// list all keyspaces
+GET /keyspaces
+[
+  {
+    name: "default",
+    id: "0"
+  },
+  {
+    name: "redis",
+    id: "1"
+  }
+]
+
+// add new keyspace
+POST /keyspaces
+{
+  name: "foo",
+}
+
+// delete a keyspace
+DELETE /keyspaces/{keyspace_name}
+{}
+
+// recover the latest deleted keyspace
+POST /keyspaces?action=flashback
+{
+  new_name: "bar",
+}
+```
+
+The keyspaces are stored in etcd and has the no limitation on the id number.
+
+1. Adding keyspace: newly added keyspaces can only be viable to new clients.
+2. Deleting keyspace:
+  2.1. Deleting keyspace only marks the metadata to inviable in PD, the data in the keyspace and metadata in PD will not be deleted automatically. Garbage collecting the data in deleted keyspaces may be introduced in the future since it increases the complexity of this RFC. To ensure no data is left, the user should clean up the keyspace before deleting the keyspace at present.
+  2.2. Every client syncs the keyspace information with the PD leader every 5 minutes. When a keyspace is deleted, the client should be aware of the deletion in 5 minutes. The keyspace can not be accessed by any living client after 5 minutes.
```

Comment:
How about make it configurable but deny extreme settings, 0 minute for example.

Reply:
This is a big topic. The main problem is that it's hard to synchronize this setting across all components including the clients the servers and pd.

Reply:
we can hardcode a setting at present. optimize the mechanism to improve user experience in the future.